### ISSUE TITLE
PC-98: FDD images with 128-byte/sector boot sectors eventually jump to E800:0002. What's that about?

### ISSUE THREAD

**Issue by joncampbell123** [joncampbell123/dosbox-x#732]
Every FDD disk image I have, where track 0 head 0 is 128 bytes/sector and the rest of the disk is 256 bytes/sector, appears to load data from disk into memory, and then RETF to E800:0002.

I'm guessing that E800:0002 might be N88 ROM BASIC, but I'm not sure.

@yksoft1 Any comments?

**Comment by joncampbell123** [comment]
There's code in Neko Project II that suggests the BIOS starts at E800:0000 and is 96KB in size.

**Comment by yksoft1** [comment]
According to my knowledge PC-98 BIOS had 8 banks, the bootloader (ITF) (BANK4) is loaded into F800:0000 and the N88-BASIC (BANK5) is loaded into E800:0000.

**Comment by yksoft1** [comment]
And there is another FDD format that may look strange to you. 256b/sector in track 0 and 1024b/sector in other tracks. For example, Nihon Falcom's earlier 2HD format.

**Comment by yksoft1** [comment]
I think there might be bugs in image format convertor (VFIC) I mainly use. Is there any tool that can make FDD images from real floppies? 
However if you use VFIC to convert those images to .d88 format, they will boot in Neko Project II. Neko Project II natively support .d88 floppy images.

BTW: Anyone interested to port D88 disk image format support from Neko Project II?

**Comment by joncampbell123** [comment]
I'm not aware of any tool, but I've been considering writing a tool in the DOSLIB project to write that format since that project now has a low-level disk reader program (for IBM PC/AT and compatibles, so far). Assuming the FDD format can represent either PC-98 or IBM formats, FDD would be a good fit for recording various IBM floppies I have with games that use weird sector layouts and/or bad sectors for copy protection.

**Comment by joncampbell123** [comment]
Eventually.

I tried loading a PC-9821 ROM BIOS to get N88 BASIC and it seems like all the code is mixed together that makes it difficult to do. I got ROM BASIC to get as far as it's prompt before crashing.

**Comment by joncampbell123** [comment]
I forgot to mention, the last time I tried it, N88 BASIC taken from a PC-9821 BIOS image apparently likes to make full use of the HMA area, meaning it will unmask the A20 line and execute most of itself from segment FFFF

**Comment by joncampbell123** [comment]
@yksoft1 I haven't yet gotten N88 BASIC to run, but I just added .D88 image support using struct typedefs copied from Neko Project II. I could not find anything on D88 from Google or DuckDuckGo so I had to use the typedefs in NP2 which seem to work well.

I'm well aware that D88 is typically for PC-88 games (which use the Z80) but I also understand from your videos that there is a PC-88 emulator for PC-98 that lets you run PC-88 games, so this should help with that.

**Comment by yksoft1** [comment]
PC-88 didn't have a floppy drive that support 2DD/2HD disks until end of life and almost nothing in PC-88 use more than 2D disks. 
Why just disable support for smaller D88 files and D88 files with multiple disks stored for now?

**Comment by yksoft1** [comment]
@joncampbell123 By the way, old MinGW (not mingw-w64) doesn't have strcasestr.

**Comment by joncampbell123** [comment]
@yksoft1 Agreed. DOSBox-X does not yet interpret the fd type field, but that could be exposed somehow to the BOOT command to reject the 2D format.

**Comment by joncampbell123** [comment]
@yksoft1 In the meantime, I'd like to know if this improves your use with the PC-88 emulator in PC-98 mode you showed in one YouTube video.

**Comment by yksoft1** [comment]
I don't think D88 support had any relation with P88SR as neither PC-88 use MS-DOS format disks at all, nor P88SR boots from real floppy drives.
However now if you map VF keys in mapper, P88SR is almost fully working in Dosbox-X.

**Comment by joncampbell123** [comment]
I just found some D88 images with PC-98 games, and so far, it's working fine. I updated the BOOT command to allow booting D88 unless the fd_type field suggests PC-88 software.

**Comment by joncampbell123** [comment]
There's usually one extra check they do with the BIOS that prevents it.

Some code was committed yesterday that should allow MEGDOS based PC-98 games to boot now.

MEGDOS it seems likes to poll the sectors going by the head using INT 1Bh before attempting to read the disk. If INT 1Bh doesn't show any movement or the sectors IDs it expects, it won't try to read the disk. Yesterday's commit added code to fake that.

**Comment by yksoft1** [comment]
@joncampbell123 Would you try those non-N88-BASIC special disk format (such as Nihon Falcom's 256/1024 byte sector format)?

**Comment by joncampbell123** [comment]
I have tried those FDD images. DOSBox-X identifies as 1024 bytes/sector and tries to boot the 1024 bytes/sector boot sector, which is zeros.

At that point, if the FDD image is broken to have two sectors per sector ID on a track, it should be fixed with a tool instead of adding more hacks.

**Comment by yksoft1** [comment]
I would upload those disk images in D88 format for you. Those FDD images are converted from NFD or D88 format using VFIC, and such converter might itself had some bugs.
Here is those image in their original distributed format.
[falcom-d88.zip](https://github.com/joncampbell123/dosbox-x/files/2042317/falcom-d88.zip)
Will remove after @joncampbell123 reply.

**Comment by yksoft1** [comment]
@joncampbell123 I'd found another strange format. 2DD disk with CHS 80/2/16 256 bytes/sector instead of CHS 80/2/8 512 bytes/sector.
One Nihon Falcom game (Sorcerian and all scenario expansions) uses this.
